### ISSUE TITLE
Port to GTK3 and new VTE

### ISSUE THREAD

**Issue by egmontkob** [asbru-cm/asbru-cm#176]
Ásbrú should be ported to GTK3. You guys all know it well, e.g. https://github.com/asbru-cm/asbru-cm/projects/1, the `feature-gtk3` branch, etc...

It's mostly not about GTK or GNOME 2/3 per se (although that too, e.g. the continuously increasing maintenance burden like in #114), but probably most importantly about VTE. GTK2 can only use VTE version 0.28 which has been unmaintained for 8+ years, and is nowadays the cause of serious terminal emulation troubles like #49, #143. There have been hundreds of other bugfixes, including a security/privacy issue of [scrollback written to disk unencrypted](http://www.climagic.com/bugreports/libvte-scrollback-written-to-disk.html), many crash fixes, tons of emulation fixes, huge speed improvement, and so on, and also many many cool new features added to GTK3-based VTE which is currently at version 0.58.

Old VTE will probably be removed by distributions in the near future: [Debian](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=885563), [Ubuntu](https://bugs.launchpad.net/ubuntu/+source/vte/+bug/1829377), [Fedora](https://bugzilla.redhat.com/show_bug.cgi?id=1739461).

---

Here's my work on porting Ásbrú to GTK3. IMO it's in pretty good state, and is quite usable (unlike the feature-gtk3 branch). I haven't checked all the features of Ásbrú (I'm not even familiar with them), but the main UI, preferences, defining a connection etc. seem to work fine, including launching a local terminal using new VTE.

The feature-gtk3 branch comments out lots of things that I didn't need to comment out, and is behind master, so I decided to base my work on master. The attached patches apply on current master (158c2abdc7). That being said, certain things I've taken from the feature-gtk3 branch (to make it easier to compare the two, pick fixes from one to the other).

- [0001-search-n-replace-Gtk2-Gtk3.patch.txt](https://github.com/asbru-cm/asbru-cm/files/3646187/0001-search-n-replace-Gtk2-Gtk3.patch.txt) – This is just an automatic search'n'replace from "Gtk2" to "Gtk3" and "GTK2" to "GTK".

- [0002-asbru.glade-from-feature-gtk3.patch.txt](https://github.com/asbru-cm/asbru-cm/files/3646189/0002-asbru.glade-from-feature-gtk3.patch.txt) – This one adds the new glade file, copied from the feature-gtk3 branch.

- [0003-asbru.glade-updates-d3b14a5089.158c2abdc7.patch.txt](https://github.com/asbru-cm/asbru-cm/files/3646193/0003-asbru.glade-updates-d3b14a5089.158c2abdc7.patch.txt) – This patch updates the glade file, by backporting the changes that happened in master since the fork.

- [0004-My-stuff.patch.txt](https://github.com/asbru-cm/asbru-cm/files/3646200/0004-My-stuff.patch.txt) – On top of the previous patches that "prepare" for GTK3, this is the real porting.

- [0005-Drag-n-drop.patch.txt](https://github.com/asbru-cm/asbru-cm/files/3651217/0005-Drag-n-drop.patch.txt) – Fix DND and a few related bits.

- [0006-Drag-n-drop-tooltip-border.patch.txt](https://github.com/asbru-cm/asbru-cm/files/3651576/0006-Drag-n-drop-tooltip-border.patch.txt) – Add a black border to DND tooltip.

- [0007-TreeView-fix-getting-node_uuid.patch.txt](https://github.com/asbru-cm/asbru-cm/files/3657689/0007-TreeView-fix-getting-node_uuid.patch.txt) – Fix finding an item's children in TreeView.

- [0008-Silly-typo.patch.txt](https://github.com/asbru-cm/asbru-cm/files/3657878/0008-Silly-typo.patch.txt) – Typo fix (and also the reason why I hate script languages as opposed to compiled ones :)).

---

Known issues, bugs, missing bits:

- Need to [migrate from libunique to GApplication or GtkApplication](https://developer.gnome.org/gtk3/stable/gtk-migrating-unique-GtkApplication.html), from `send_message()` to probably GApplication's "command-line" signal and friends.

- Remembering the history of terminal commands doesn't work (I've commented it out). VTE's `get_text_range()` gives this error: "gperl-i11n-invoke-c.c:582:_allocate_out_mem: assertion failed: (interface_info)". Looks like a nasty issue with VTE's annotation or the introspection framework. Might need help from introspection and Perl binding experts.

- ~Port drag and drop.~ Update: It's ported now, but see a comment below.

- Port tray icon support (haven't tested).

- Make sure that X11 compatibility layer is used even when on Wayland, rather than being a native Wayland app, so that embedding an external window based on XID still works; or at least make sure it degrades gracefully on Wayland.

- Update the documentation, user facing messages, deb/rpm definitions etc.

- Testing, testing, testing. Walking through all the features. As soon as something doesn't work, or an error is printed, track it down and fix it.

- Update for new stuff in VTE:

  - Transparency needs to be done differently. `set_color_background()` needs to get a color with an alpha channel, plus the widget needs to be app_paintable. Maybe something more is needed too, I'm not sure. See VTE-based apps that support transparency for reference, e.g. Tilix, Xfce4-terminal, Guake, and a Fedora/Ubuntu patch for GNOME Terminal. Or just remove this feature.

  - `set_emulation()` is no longer available, VTE can no longer try to emulate various behaviors as per their terminfo definitions, the emulation behavior is hardwired. Remove the corresponding config option and UI field.

  - `set_visible_bell()` is no longer available. You might remove this option, or implement showing some custom UI element (e.g. a bell symbol) on the bell signal.

  - `set_bold_is_bright()` is new in VTE 0.52, and defaults to being disabled since 0.56. This latter one results in darker colors that not everyone likes, for compatibility it would be great to add a config option. ([Discussions](https://bugzilla.gnome.org/show_bug.cgi?id=762247))

  - New VTE lets you display the right-click context menu only if the application running inside is not interested in mouse events, and pass it to the application otherwise. ([PAC bug](https://github.com/perseo22/pacmanager/issues/55), [VTE fix](https://bugzilla.gnome.org/show_bug.cgi?id=13299))

  - You might go through the API docs and decide what other new stuff to add. E.g. cursor blinking, text blinking, mouse pointer autohide...

---

I sincerely hope that my work gives a boost to porting to GTK3 and that it will happen in the foreseeable future.

(Disclaimer: I don't use Ásbrú myself and I'm not particularly interested in this project. What I'm interested in is terminal emulation in general, and retiring old VTE in favor of the new one ASAP.)

**Comment by gfrenoy** [comment]
Waow!  Great contribution, thanks a million @egmontkob :-)

I did not find time today to review this in detail and give it a try but I'll probably delete (or rename) the ```feature-gtk3``` and re-start this from your work.

Especially if we can have something that is functional.  This a stage that I never reached so far :-/

It would indeed be a tremendous objective to reach if we can get rid of that legacy VTE build !

Is there anything we can do to have a bit of your time and expertise to help me (us) on this ?  Would you mind joining us on [IRC](https://webchat.st-city.net/#asbru-cm) or [Matrix](https://matrix.to/#/#asbru-cm:scificenter.de) (see https://www.asbru-cm.net/support.html) so we can discuss this further ?

Thanks again for this great post, it really made my day !

**Comment by egmontkob** [comment]
Hi,

> I'll probably delete (or rename) the `feature-gtk3`

Please don't remove it, at least not until gtk3 port is considered "ready", merged into master and released as stable. It might contain precious pieces of work that should be merged with mine.

> Would you mind joining us on IRC [...]

I generally prefer non-real-time, as well as archived communication (e.g. commenting here). We could arrange an IRC session, but let's do if after you take a close look at the current state, know what questions you have, where you think you'd need my help, and if these are questions that are not really suitable for github discussion (e.g. require more interactivity).

**Comment by egmontkob** [comment]
> VTE's `get_text_range()` gives this error: "gperl-i11n-invoke-c.c:582:_allocate_out_mem: assertion failed: (interface_info)".

I played a bit with this one, and I think I give it up here.

The same method is used successfully in Terminator's logging plugin (Python).

If I alter the `(out caller-allocates)` annotation to simply `(out)` in VTE's source, the error message becomes a segfault instead.

If I remove `(out ...)` altogether (i.e. make it an `in` parameter), make it `(allow-none)` and pass it an `undef` (as we really don't care about this parameter) then it works like a charm.

Possibilities I see:

- Consult with VTE's main developer (he knows much more about these stuff than I do) whether our annotations are really good here, and whether we can (and should) change them to please the Perl binding. Mind you, I'm really not sure why it's an `out` parameter when its NULL-ness (undef-ness) actually carries information to the method (it saves work and memory if it's NULL/undef); shouldn't it be `in` or `inout` instead?

- Consult with Introspection's developers about this setup (`GArray`, which I believe is translated to native Perl array, as `(out ...)`) possibly not being supported and whether they're up to adding/fixing it.

- Write a small wrapper in C (and ship it as part of Ásbrú) that just skips this problematic parameter, and create a Perl binding for that that we could use.

- Drop this weird hack of Ásbrú of supporting command line history on its own.

**Comment by egmontkob** [comment]
DND is now ported, see the updated opening comment.

When you drag a connection definition from the left-side treeview, the tooltip isn't as prominent as with GTK2, due to the lack of a black border. This is because I altered the code to work with a regular GtkLabel, rather than manually painting the text on a canvas.

You can experiment with adding a GtkFrame around it, or finding other means of drawing a border (e.g. placing it inside a GtKDrawingArea, not sure if that would work). Alternatively, you can go back to the previous model of drawing manually, the original code then needs to be ported from GdkPixbuf to Cairo. Should be a pretty straightforward task if you prefer the old look and prefer the code to remain more similar to the GTK2 one. I'm not planning to further work on this.

**Comment by gfrenoy** [comment]
Thanks again for your tremendous contribution @egmontkob !

I created branch [gtk3](https://github.com/asbru-cm/asbru-cm/tree/gtk3) with all your patches and merged back the latest changes from master.

It would be great if we can continue this effort on that branch until we release a last version on Gtk2 and then merge back gtk3 to the master branch.

**Comment by egmontkob** [comment]
Attached a 7th patch fixing #184 and #186. I incorrectly assumed that a missing last parameter to `get_value` (which is no longer permitted in GTK3) defaulted 0, whereas it apparently defaulted to the last column.

**Comment by egmontkob** [comment]
I sneaked in an 8th patch the other day, without making a comment about it. Could you please apply that? (I promise I won't sneak-edit comments this way again.)

**Comment by gfrenoy** [comment]
What a journey since you gave that "kick" into the project !  The work is not yet fully completed but let's close this issue for now, we have more accurate description in the other issues.
